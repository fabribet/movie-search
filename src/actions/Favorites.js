export const types = {
  ADD_FAVORITE: 'ADD_FAVORITE',
  REMOVE_FAVORITE: 'REMOVE_FAVORITE'
}

export const actions = {
  /**
   * Generates an Add Favorite Action
   */
  AddFavorite (movie) {
    return {
      type: types.ADD_FAVORITE,
      movie
    }
  },

  /**
   * Generates a Remove Favorite Action
   */
  RemoveFavorite (movieId) {
    return {
      type: types.REMOVE_FAVORITE,
      movieId
    }
  }
}

/**
 * Returns the selectors to access the Favorites part of the state.
 */
export const selectors = {
  GetFavorites: state => state.Favorites
}
