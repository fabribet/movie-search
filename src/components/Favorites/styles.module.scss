.favorites {
  height: 100px;
  width: 90%;
  max-width: 1000px;
  margin: auto;

  .title {
    display: flex;
    margin: 10px auto;
    width: fit-content;

    .favoriteImg {
      margin-left: 3px;
      height: 12px;
      width: 12px;
    }
  }

  .container {
    display: flex;
    padding: 5px 10px 5px;
    font-size: 10px;
    overflow-x: scroll;
    height: 70px;
    &.empty {
      margin: auto;
      text-align: center;
      width: fit-content;
    }
  }
}
