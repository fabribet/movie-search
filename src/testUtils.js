import configureStore from 'redux-mock-store'

// Mock the canvas usage for external library usage on tests.
export function mockCanvas () {
  const createElement = document.createElement.bind(document)
  document.createElement = tagName => {
    const element = createElement(tagName)
    if (tagName === 'canvas') {
      element.getContext = () => ({})
    }
    return element
  }
}

/**
 * Create Mock store for testing
 */
const mockStore = configureStore([])
export function createMockStore (initialState = {
  Favorites: {
    favorites: {}
  },
  TMDB: {
    apiConfig: {
      loading: false,
      error: null,
      data: {
        images: {
          base_url: ''
        }
      }
    }
  }
}) {
  return mockStore(initialState)
}

export const FAVORITES_DATA = {
  1: {
    id: 1,
    title: 'Movie 1',
    poster_path: '/asdasdas.jpg'
  },

  2: {
    id: 2,
    title: 'Movie 2',
    poster_path: '/asdasdas2.jpg'
  }
}
