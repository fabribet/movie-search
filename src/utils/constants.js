
/**
 * Text displayed when no favorites have been added
 */
export const EMPTY_FAVORITES_TXT = 'Add favorites to your list to see them displayed here.'

/**
 * Text displayed when no movies have been found
 */
export const EMPTY_MOVIE_LIST_TXT = 'There are no movies that match the criteria, try again.'
