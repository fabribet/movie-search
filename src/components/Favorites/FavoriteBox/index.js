import { connect } from 'react-redux'
import { actions } from '../../../actions/Favorites'
import FavoriteBox from './FavoriteBox'

const mapStateToProps = state => ({
  // If the API configuration data has been obtained the Images base url is passed.
  imgUrl: state.TMDB.apiConfig.data ? state.TMDB.apiConfig.data.images.base_url : null
})

const mapDispatchToProps = dispatch => ({
  removeFavorite (movieId) {
    dispatch(actions.RemoveFavorite(movieId))
  }
})

export default connect(mapStateToProps, mapDispatchToProps)(FavoriteBox)
