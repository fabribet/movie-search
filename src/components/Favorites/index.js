import { connect } from 'react-redux'
import Favorites from './Favorites'

const mapStateToProps = state => {
  return {
    favorites: state.Favorites.favorites
  }
}

export default connect(mapStateToProps)(Favorites)
