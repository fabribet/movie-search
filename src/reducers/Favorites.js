import { types } from '../actions/Favorites'

export default (state = { favorites: {} }, action) => {
  switch (action.type) {
    case types.ADD_FAVORITE:
      return {
        ...state,
        favorites: {
          ...state.favorites,
          [action.movie.id]: action.movie
        }
      }

    case types.REMOVE_FAVORITE:
      const favorites = Object.assign({}, state.favorites)
      if (favorites[action.movieId]) delete favorites[action.movieId]
      return {
        ...state,
        favorites
      }

    default:
      return state
  }
}
