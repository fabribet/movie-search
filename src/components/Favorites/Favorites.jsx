import React from 'react'
import PropTypes from 'prop-types'
import styles from './styles.module.scss'

import heartImg from './imgs/heart-solid.svg'
import { EMPTY_FAVORITES_TXT } from '../../utils/constants'
import FavoriteBox from './FavoriteBox'

/**
 * Favorites - React component.
 * Renders a list of favorited movies.
 *
 * Properties
 * - favorites {Object} - The list of favorites (as an object).
 */
function Favorites (props) {
  // Favorites are stored as a key<movieId => value<movieObject> pair in an object.
  const favoritesKeys = Object.keys(props.favorites)
  return (
    <div className={styles.favorites}>
      <h5 className={styles.title}>Favorite Movies <img className={styles.favoriteImg} src={heartImg} alt={'favorite'}/></h5>
      <div className={`${styles.container} ${!favoritesKeys || !favoritesKeys.length ? styles.empty : ''}`}>
        {favoritesKeys && favoritesKeys.length
          ? favoritesKeys.map((key) => <FavoriteBox key={key} movie={props.favorites[key]} />)
          : EMPTY_FAVORITES_TXT
        }
      </div>
    </div>
  )
}

Favorites.propTypes = {
  favorites: PropTypes.shape({}).isRequired
}

export default Favorites
