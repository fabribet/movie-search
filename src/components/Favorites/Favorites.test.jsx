import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'

import Favorites from './Favorites'
import { EMPTY_FAVORITES_TXT } from '../../utils/constants'
import { mockCanvas, createMockStore, FAVORITES_DATA } from './../../testUtils'

mockCanvas()
const store = createMockStore()

const favorites =
  <Provider store={store}>
    <Favorites favorites={FAVORITES_DATA} />
  </Provider>

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(
    favorites,
    div
  )
})

it('Renders favorites correctly', () => {
  const tree = renderer
    .create(favorites)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('Renders an empty list message when no favorites have been added', () => {
  const { queryByText } = render(
    <Provider store={store}>
      <Favorites favorites={{}} />
    </Provider>
  )

  expect(queryByText(EMPTY_FAVORITES_TXT)).not.toBeNull()
})

it('Renders the list and no text is displayed', () => {
  const { queryByText } = render(favorites)

  expect(queryByText(EMPTY_FAVORITES_TXT)).toBeNull()
})
