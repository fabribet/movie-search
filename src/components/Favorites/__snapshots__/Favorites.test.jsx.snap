// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders favorites correctly 1`] = `
<div
  className="favorites"
>
  <h5
    className="title"
  >
    Favorite Movies 
    <img
      alt="favorite"
      className="favoriteImg"
      src="heart-solid.svg"
    />
  </h5>
  <div
    className="container "
  >
    <div
      className="box"
      onClick={[Function]}
    >
      <div
        className="img"
        style={
          Object {
            "backgroundImage": "url(movie-poster.png)",
          }
        }
      />
      <div
        className="movieInfo"
      >
        <div
          className="titleWrapper"
        >
          <h4
            className="title"
          >
            <span
              width={0}
            >
              <span />
              <span>
                Movie 1
              </span>
              <span
                style={
                  Object {
                    "left": 0,
                    "position": "fixed",
                    "top": 0,
                    "visibility": "hidden",
                  }
                }
              >
                ...
              </span>
            </span>
          </h4>
          <div
            className="removeBtn"
            id="fb-removebtn"
            onClick={[Function]}
            title="Remove favorite"
          />
        </div>
      </div>
    </div>
    <div
      className="box"
      onClick={[Function]}
    >
      <div
        className="img"
        style={
          Object {
            "backgroundImage": "url(movie-poster.png)",
          }
        }
      />
      <div
        className="movieInfo"
      >
        <div
          className="titleWrapper"
        >
          <h4
            className="title"
          >
            <span
              width={0}
            >
              <span />
              <span>
                Movie 2
              </span>
              <span
                style={
                  Object {
                    "left": 0,
                    "position": "fixed",
                    "top": 0,
                    "visibility": "hidden",
                  }
                }
              >
                ...
              </span>
            </span>
          </h4>
          <div
            className="removeBtn"
            id="fb-removebtn"
            onClick={[Function]}
            title="Remove favorite"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
